pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_conversion.go

```diff
@@ -210,42 +211,57 @@ func (r *GatewayReconciler) gapiToKumaGateway(
 			listener.Hostname = string(*l.Hostname)
 		}
 
-		var unresolvableCertRefs []string
+		type badCertRef struct {
+			message string
+			reason  string
+		}
+		var unresolvableCertRef *badCertRef
 		if l.TLS != nil {
+			var referencedSecrets []kube_types.NamespacedName
 			for _, certRef := range l.TLS.CertificateRefs {
 				policyRef := policy.PolicyReferenceSecret(policy.FromGatewayIn(gateway.Namespace), certRef)
 
-				permitted, err := policy.IsReferencePermitted(ctx, r.Client, policyRef)
-				if err != nil {
+				if permitted, err := policy.IsReferencePermitted(ctx, r.Client, policyRef); err != nil {
 					return nil, nil, err
+				} else if !permitted {
+					name := fmt.Sprintf("%q %q", policyRef.GroupKindReferredTo().String(), policyRef.NamespacedNameReferredTo().String())
+					unresolvableCertRef = &badCertRef{
+						reason:  string(gatewayapi.ListenerReasonRefNotPermitted),
+						message: fmt.Sprintf("reference to %s not permitted by any ReferenceGrant", name),
+					}
+					break
 				}
 
-				if !permitted {
-					message := fmt.Sprintf("%q %q", policyRef.GroupKindReferredTo().String(), policyRef.NamespacedNameReferredTo().String())
-					unresolvableCertRefs = append(unresolvableCertRefs, message)
+				if err := r.Client.Get(ctx, policyRef.NamespacedNameReferredTo(), &kube_core.Secret{}); err != nil {
+					if kube_apierrs.IsNotFound(err) {
+						name := fmt.Sprintf("%q %q", policyRef.GroupKindReferredTo().String(), policyRef.NamespacedNameReferredTo().String())
+						unresolvableCertRef = &badCertRef{
+							reason:  string(gatewayapi.ListenerReasonInvalidCertificateRef),
+							message: fmt.Sprintf("invalid reference to %s", name),
+						}
+						break
+					} else {
+						return nil, nil, err
+					}
```

Comment:
I don't think there is need to have an else branch here
```suggestion
					}
					
					return nil, nil, err
```